tests/subsys/usb/desc_sections/src/desc_sections.c

```diff
@@ -198,41 +198,49 @@ static void check_endpoint_allocation(struct usb_desc_header *head)
 	}
 }
 
+/* Determine the number of bytes spanned between two linker-defined
+ * symbols that are normally interpreted as pointers.
+ *
+ * Note that these symbols provide the address of packed structures,
+ * so alignment restrictions are not relevant.
```

Comment:
I find this comment confusing. Packing structures reduces alignment of its members but not the alignment of the structure itself. Which is it here?

Reply:
Ok, let's dive deeper into the rabbit hole, and maybe start digging a subbasement.

Consider:
```
typedef struct p64 {
  uint64_t u64;
  uint8_t u8;
} __attribute__((__packed__)) p64;

p64 arr[1];
p64 *ap;
```

The size of the structure is 9 bytes (if byte=char).  The size of the array is 9.  Any value of `ap` has no clearly defined alignment, because it might be into an array of any number of instances.

What the comment was meant to convey is that there's no point in checking the alignment of the symbols (which are pointers) because there isn't any.

Now that's clearly true for the end symbol, because `arr+1` which is a valid pointer that is 9 bytes past `arr`.  The end symbol in the test can't meet the struct alignment requirements because then the calculation of the array length (which is 133 bytes for the descriptor sequence in the test) would be wrong.

You may well be right that there's a requirement that the first structure has to be placed to guarantee proper alignment of its `u64` member.  I don't think that's worth checking because it's  important only if the compiler happens to require aligned access when it knows it's accessing the first member of the first element of an array.

Also, I don't know how to do that check, since I have no idea how to portably determine the alignment requirements of a packed struct type at compile time.   At least not in C.

So: Would you like me to remove that sentence from the comment?

Reply:
> So: Would you like me to remove that sentence from the comment? 

Please. All that "Note.. relevant." sentence

Reply:
Done.

Reply:
> You may well be right that there's a requirement that the first structure has to be placed to guarantee proper alignment of its u64 member. 

I was wrong actually :-) Packing a struct is equivalent to aligned(1) on every member. This indirectly affects the whole structure because the alignment of the whole structure is always defined by the "largest" alignment across all its members: in this case 1.